twoliter/embedded/imghelper

```diff
@@ -108,6 +110,228 @@ mkfs_data_xfs() {
+    -in "${SBKEYS}/code-sign.crt" \
+    -certfile "${SBKEYS}/vendor.crt" \
+    -out "${SBKEYS}/code-sign.p12"
+
+  # Import certificates and private key archive.
+  local pedb="/etc/pki/pesign"
+
+  certutil -d "${pedb}" -A -n db -i "${SBKEYS}/db.crt" -t ",,C"
+  certutil -d "${pedb}" -A -n shim-sign-key -i "${SBKEYS}/shim-sign.crt" -t ",,P"
+  pk12util -d "${pedb}" -i "${SBKEYS}/shim-sign.p12" -W ""
+
+  certutil -d "${pedb}" -A -n vendor -i "${SBKEYS}/vendor.crt" -t ",,C"
+  certutil -d "${pedb}" -A -n code-sign-key -i "${SBKEYS}/code-sign.crt" -t ",,P"
+  pk12util -d "${pedb}" -i "${SBKEYS}/code-sign.p12" -W ""
+
+  certutil -d "${pedb}" -L
+  shim_sign_key=(-c shim-sign-key)
+  code_sign_key=(-c code-sign-key)
+
+  sbsetup_wrapup local
+}
+
+uefi_pesign() {
+  local efi_mount arch shims shim grubs grub mokms mokm
+  local -n shim_sign_key code_sign_key
+  efi_mount="${1:?}"
+  arch="${2:?}"
+  shim_sign_key="${3:?}"
+  code_sign_key="${4:?}"
+
+  pushd "${efi_mount}/EFI/BOOT" >/dev/null || exit 1
+
+  shims=(boot*.efi)
+  shim="${shims[0]}"
+  grubs=(grub*.efi)
+  grub="${grubs[0]}"
+  mokms=(mm*.efi)
+  mokm="${mokms[0]}"
+
+  # Convert the vendor certificate to the expected format.
+  cert_table "${SBKEYS}/vendor.cer" "${SBKEYS}/vendor.obj"
+
+  # Replace the embedded vendor certificate, then sign shim with the db key.
+  if [[ "$(pesign -i "${shim}" -l)" == "No signatures found." ]]; then
+    mv "${shim}" "${shim}.unsigned"
+  else
+    pesign -i "${shim}" -o "${shim}.unsigned" -u 0 -r
+  fi
+  "${arch}-objdumpcopy" "${shim}.unsigned" \
+    --update-section ".vendor_cert=${SBKEYS}/vendor.obj"
+  pesign -i "${shim}.unsigned" -o "${shim}.signed" -s "${shim_sign_key[@]}"
+  mv "${shim}.signed" "${shim}"
+  rm "${shim}.unsigned"
+  pesign -i "${shim}" -l
+  pesigcheck -i "${shim}" -n 0 -c "${SBKEYS}/db.cer"
+
+  # Remove the original key if necessary, then sign the MOK manager as well.
+  if [[ "$(pesign -i "${mokm}" -l)" == "No signatures found." ]]; then
+    mv "${mokm}" "${mokm}.unsigned"
+  else
+    pesign -i "${mokm}" -o "${mokm}.unsigned" -u 0 -r
+  fi
+  pesign -i "${mokm}.unsigned" -o "${mokm}.signed" -s "${code_sign_key[@]}"
+  mv "${mokm}.signed" "${mokm}"
+  rm "${mokm}.unsigned"
+  pesign -i "${mokm}" -l
+  pesigcheck -i "${mokm}" -n 0 -c "${SBKEYS}/vendor.cer"
+
+  # Remove the original vendor key if necessary, replace the embedded gpg
+  # public key, then sign grub with the vendor key.
+  if [[ "$(pesign -i "${grub}" -l)" == "No signatures found." ]]; then
+    mv "${grub}" "${grub}.unsigned"
+  else
+    pesign -i "${grub}" -o "${grub}.unsigned" -u 0 -r
+  fi
+  "${arch}-objdumpcopy" "${grub}.unsigned" \
+    --file-alignment 4096 \
+    --update-section ".pubkey=${SBKEYS}/config-sign.pubkey"
+  pesign -i "${grub}.unsigned" -o "${grub}.signed" -s "${code_sign_key[@]}"
+  mv "${grub}.signed" "${grub}"
+  rm "${grub}.unsigned"
+  pesign -i "${grub}" -l
+  pesigcheck -i "${grub}" -n 0 -c "${SBKEYS}/vendor.cer"
```

Comment:
As @rpkelly mentioned, the duplication here is a sign that this function is doing too much, which is part of why its signature is inconsistent.